File: code/backend/python/app/adapters/persistence/match/turn_cycle_store_adapter.py
"""Step 24 — SQLAlchemy adapter implementing :class:`TurnCycleStorePort`."""
import uuid as uuid_lib
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional

from app.adapters.persistence.auth.models import User
from app.adapters.persistence.match.models import (
    GamingCharacterInstanceEntity,
    GamingMatchEntity,
    GamingTurnQueueEntity,
)
from app.core.ports.match.turn_ports import TurnCycleStorePort


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


def _new_uuid() -> str:
    return str(uuid_lib.uuid4())


class TurnCycleStoreAdapter(TurnCycleStorePort):
    def __init__(self, session_factory):
        self.session_factory = session_factory

    def find_match_by_uuid(self, uuid: str) -> Optional[Dict[str, Any]]:
        if not uuid:
            return None
        with self.session_factory() as session:
            m = session.query(GamingMatchEntity).filter(GamingMatchEntity.uuid == uuid).first()
            if m is None:
                return None
            return {
                "id": m.id,
                "uuid": m.uuid,
                "status": m.status,
                "current_clock": m.current_clock or 0,
                "id_user_creator": m.id_user_creator,
                "id_character_current_turn": m.id_character_current_turn,
            }

    def find_characters_by_match_id(self, id_match: int) -> List[Dict[str, Any]]:
        with self.session_factory() as session:
            rows = (
                session.query(GamingCharacterInstanceEntity)
                .filter(GamingCharacterInstanceEntity.id_match == id_match)
                .all()
            )
            return [
                {
                    "id": c.id,
                    "uuid": c.uuid,
                    "id_user": c.id_user,
                    "dexterity": c.dexterity,
                    "intelligence": c.intelligence,
                    "constitution": c.constitution,
                    "life": c.life,
                }
                for c in rows
            ]

    def replace_queue(self, id_match: int, rows: List[Dict[str, Any]]) -> None:
        with self.session_factory() as session:
            session.query(GamingTurnQueueEntity).filter(
                GamingTurnQueueEntity.id_match == id_match
            ).delete()
            now = _now_iso()
            for r in rows:
                session.add(GamingTurnQueueEntity(
                    id_match=id_match,
                    id_character_match=r["id_character_match"],
                    uuid=_new_uuid(),
                    clock=r["clock"],
                    timestamp_start=r.get("timestamp_start"),
                    timestamp_end=r.get("timestamp_end"),
                    pass_counter=r.get("pass_counter", 0),
                    priority=r["priority"],
                    status=r["status"],
                    ts_insert=now,
                    ts_update=now,
                ))
            session.commit()

    def find_queue_by_match_id(self, id_match: int) -> List[Dict[str, Any]]:
        with self.session_factory() as session:
            rows = (
                session.query(GamingTurnQueueEntity)
                .filter(GamingTurnQueueEntity.id_match == id_match)
                .order_by(GamingTurnQueueEntity.priority.desc())
                .all()
            )
            return [
                {
                    "id_character_match": r.id_character_match,
                    "uuid": r.uuid,
                    "clock": r.clock,
                    "priority": r.priority,
                    "status": r.status,
                    "pass_counter": r.pass_counter,
                    "timestamp_start": r.timestamp_start,
                    "timestamp_end": r.timestamp_end,
                }
                for r in rows
            ]

    def save_queue_row(self, id_match: int, row: Dict[str, Any]) -> None:
        with self.session_factory() as session:
            entity = (
                session.query(GamingTurnQueueEntity)
                .filter(
                    GamingTurnQueueEntity.id_match == id_match,
                    GamingTurnQueueEntity.id_character_match == row["id_character_match"],
                )
                .first()
            )
            if entity is None:
                return
            entity.status = row["status"]
            entity.pass_counter = row.get("pass_counter", entity.pass_counter)
            entity.timestamp_start = row.get("timestamp_start")
            entity.timestamp_end = row.get("timestamp_end")
            entity.ts_update = _now_iso()
            session.commit()

    def update_match_status_and_turn(self, id_match: int, status: str,
                                     id_character_current_turn: Optional[int]) -> None:
        with self.session_factory() as session:
            m = session.query(GamingMatchEntity).filter(GamingMatchEntity.id == id_match).first()
            if m is None:
                return
            m.status = status
            m.id_character_current_turn = id_character_current_turn
            m.ts_update = _now_iso()
            session.commit()

    def find_user_id_by_uuid(self, user_uuid: str) -> Optional[int]:
        if not user_uuid:
            return None
        with self.session_factory() as session:
            u = session.query(User).filter(User.uuid == user_uuid).first()
            return u.id if u else None

File: code/backend/java/core/src/main/java/games/paths/core/persistence/match/TurnCycleStoreAdapter.java
package games.paths.core.persistence.match;

import games.paths.core.entity.match.GamingCharacterInstanceEntity;
import games.paths.core.entity.match.GamingCharacterInstanceEntityId;
import games.paths.core.entity.match.GamingMatchEntity;
import games.paths.core.entity.match.GamingTurnQueueEntity;
import games.paths.core.entity.match.LogClockHistoryEntity;
import games.paths.core.entity.story.StoryEntity;
import games.paths.core.entity.story.TextEntity;
import games.paths.core.port.match.TurnCycleStorePort;
import games.paths.core.repository.match.GamingCharacterInstanceRepository;
import games.paths.core.repository.match.GamingMatchRepository;
import games.paths.core.repository.match.GamingTurnQueueRepository;
import games.paths.core.entity.match.LogEventsEntity;
import games.paths.core.repository.match.LogClockHistoryRepository;
import games.paths.core.repository.match.LogEventsRepository;
import games.paths.core.repository.story.StoryRepository;
import games.paths.core.repository.story.TextRepository;
import org.springframework.stereotype.Repository;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

/**
 * TurnCycleStoreAdapter - JPA adapter implementing {@link TurnCycleStorePort}
 * over gaming_match / gaming_character_instance / gaming_turn_queue (Step 24).
 */
@Repository
@Transactional
public class TurnCycleStoreAdapter implements TurnCycleStorePort {

    private final GamingMatchRepository matchRepository;
    private final GamingCharacterInstanceRepository characterRepository;
    private final GamingTurnQueueRepository turnQueueRepository;
    private final LogClockHistoryRepository logClockHistoryRepository;
    private final LogEventsRepository logEventsRepository;
    private final StoryRepository storyRepository;
    private final TextRepository textRepository;

    public TurnCycleStoreAdapter(GamingMatchRepository matchRepository,
                                 GamingCharacterInstanceRepository characterRepository,
                                 GamingTurnQueueRepository turnQueueRepository,
                                 LogClockHistoryRepository logClockHistoryRepository,
                                 LogEventsRepository logEventsRepository,
                                 StoryRepository storyRepository,
                                 TextRepository textRepository) {
        this.matchRepository = matchRepository;
        this.characterRepository = characterRepository;
        this.turnQueueRepository = turnQueueRepository;
        this.logClockHistoryRepository = logClockHistoryRepository;
        this.logEventsRepository = logEventsRepository;
        this.storyRepository = storyRepository;
        this.textRepository = textRepository;
    }

    @Override
    @Transactional(readOnly = true)
    public Optional<MatchView> findMatchByUuid(String uuid) {
        return matchRepository.findByUuid(uuid).map(m -> new MatchView(
                m.getId(), m.getUuid(), m.getStatus(),
                m.getCurrentClock() == null ? 0 : m.getCurrentClock(),
                m.getIdUserCreator(), m.getIdCharacterCurrentTurn()));
    }

    @Override
    @Transactional(readOnly = true)
    public List<CharacterTurnView> findCharactersByMatchId(long idMatch) {
        List<CharacterTurnView> out = new ArrayList<>();
        for (GamingCharacterInstanceEntity c : characterRepository.findByIdMatch(idMatch)) {
            out.add(toView(c));
        }
        return out;
    }

    private static CharacterTurnView toView(GamingCharacterInstanceEntity c) {
        return new CharacterTurnView(
                c.getId(), c.getUuid(), c.getIdUser(),
                nz(c.getDexterity()), nz(c.getIntelligence()),
                nz(c.getConstitution()), nz(c.getLife()),
                nz(c.getEnergy()), Boolean.TRUE.equals(c.getIsSleeping()));
    }

    @Override
    public void replaceQueue(long idMatch, List<QueueRow> rows) {
        turnQueueRepository.deleteByIdMatch(idMatch);
        List<GamingTurnQueueEntity> entities = new ArrayList<>();
        for (QueueRow r : rows) {
            GamingTurnQueueEntity e = new GamingTurnQueueEntity();
            e.setIdMatch(idMatch);
            e.setIdCharacterMatch(r.idCharacterMatch());
            e.setClock(r.clock());
            e.setPriority(r.priority());
            e.setStatus(r.status());
            e.setPassCounter(r.passCounter());
            e.setTimestampStart(r.timestampStart());
            e.setTimestampEnd(r.timestampEnd());
            entities.add(e);
        }
        turnQueueRepository.saveAll(entities);
    }

    @Override
    @Transactional(readOnly = true)
    public List<QueueRow> findQueueByMatchId(long idMatch) {
        List<QueueRow> out = new ArrayList<>();
        for (GamingTurnQueueEntity e : turnQueueRepository.findByIdMatchOrderByPriorityDesc(idMatch)) {
            out.add(new QueueRow(e.getIdCharacterMatch(), e.getUuid(), nz(e.getClock()),
                    e.getPriority() == null ? 0L : e.getPriority(), e.getStatus(),
                    nz(e.getPassCounter()), e.getTimestampStart(), e.getTimestampEnd()));
        }
        return out;
    }

    @Override
    public void saveQueueRow(long idMatch, QueueRow row) {
        GamingTurnQueueEntity e = turnQueueRepository
                .findByIdMatchAndIdCharacterMatch(idMatch, row.idCharacterMatch())
                .orElse(null);
        if (e == null) return;
        e.setStatus(row.status());
        e.setPassCounter(row.passCounter());
        e.setPriority(row.priority());
        e.setClock(row.clock());
        e.setTimestampStart(row.timestampStart());
        e.setTimestampEnd(row.timestampEnd());
        turnQueueRepository.save(e);
    }

    @Override
    public void updateMatchStatusAndTurn(long idMatch, String status, Long idCharacterCurrentTurn) {
        GamingMatchEntity m = matchRepository.findById(idMatch).orElse(null);
        if (m == null) return;
        if (status != null) m.setStatus(status);
        m.setIdCharacterCurrentTurn(idCharacterCurrentTurn);
        matchRepository.save(m);
    }

    // ── Step 25: time advancement & clock cycle ─────────────────────────────

    @Override
    @Transactional(readOnly = true)
    public Optional<CharacterTurnView> findCharacterByMatchAndUser(long idMatch, long idUser) {
        return characterRepository.findByIdMatchAndIdUser(idMatch, idUser)
                .map(TurnCycleStoreAdapter::toView);
    }

    @Override
    public void setCharacterSleeping(long idMatch, long idCharacter, boolean sleeping) {
        characterRepository.findById(new GamingCharacterInstanceEntityId(idCharacter, idMatch))
                .ifPresent(c -> {
                    c.setIsSleeping(sleeping);
                    characterRepository.save(c);
                });
    }

    @Override
    public void wakeAllCharacters(long idMatch) {
        List<GamingCharacterInstanceEntity> characters = characterRepository.findByIdMatch(idMatch);
        for (GamingCharacterInstanceEntity c : characters) {
            if (Boolean.TRUE.equals(c.getIsSleeping())) {
                c.setIsSleeping(false);
            }
        }
        characterRepository.saveAll(characters);
    }

    @Override
    public void setAllCharactersSleeping(long idMatch) {
        List<GamingCharacterInstanceEntity> characters = characterRepository.findByIdMatch(idMatch);
        for (GamingCharacterInstanceEntity c : characters) {
            c.setIsSleeping(true);
        }
        characterRepository.saveAll(characters);
    }

    @Override
    public int incrementMatchClock(long idMatch) {
        GamingMatchEntity m = matchRepository.findById(idMatch).orElse(null);
        if (m == null) return 0;
        int newClock = nz(m.getCurrentClock()) + 1;
        m.setCurrentClock(newClock);
        matchRepository.save(m);
        return newClock;
    }

    @Override
    public void insertClockHistory(long idMatch, int clock) {
        LogClockHistoryEntity e = new LogClockHistoryEntity();
        e.setId(logClockHistoryRepository.findMaxId() + 1);
        e.setIdMatch(idMatch);
        e.setClock(clock);
        String now = java.time.Instant.now().toString();
        e.setTimestampStart(now);
        logClockHistoryRepository.save(e);
    }

    @Override
    public void logSleep(long idMatch, long idCharacter, int clock) {
        LogEventsEntity e = new LogEventsEntity();
        e.setId(logEventsRepository.findMaxId() + 1);
        e.setIdMatch(idMatch);
        e.setIdCharacterMatch(idCharacter);
        e.setClock(clock);
        e.setLogMessage("ACTION_SLEEP");
        logEventsRepository.save(e);
    }

    @Override
    @Transactional(readOnly = true)
    public ClockLabels findStoryClockLabels(long idMatch, String lang) {
        GamingMatchEntity m = matchRepository.findById(idMatch).orElse(null);
        if (m == null || m.getIdStory() == null) {
            return new ClockLabels(null, null);
        }
        StoryEntity story = storyRepository.findById(m.getIdStory()).orElse(null);
        if (story == null) {
            return new ClockLabels(null, null);
        }
        String singular = resolveText(story.getId(), story.getIdTextClockSingular(), lang);
        String plural = resolveText(story.getId(), story.getIdTextClockPlural(), lang);
        return new ClockLabels(singular, plural);
    }

    private String resolveText(Long idStory, Integer idText, String lang) {
        if (idStory == null || idText == null) return null;
        String effectiveLang = (lang != null && !lang.isBlank()) ? lang : "en";
        String text = firstShortText(idStory, idText, effectiveLang);
        if (text == null && !"en".equals(effectiveLang)) {
            text = firstShortText(idStory, idText, "en");
        }
        return text;
    }

    private String firstShortText(Long idStory, Integer idText, String lang) {
        List<TextEntity> texts = textRepository.findByIdStoryAndIdTextAndLang(idStory, idText, lang);
        return texts.isEmpty() ? null : texts.get(0).getShortText();
    }

    private static int nz(Integer v) {
        return v == null ? 0 : v;
    }
}
